uber/scheduler.py

```diff
@@ -0,0 +1,126 @@
+from sideboard.lib import DaemonTask, on_startup
+from uber.decorators import timed, swallow_exceptions
+from datetime import datetime
+import time
+import schedule
+from collections import defaultdict
+
+# don't start the email sending in the main thread. everything else is cool.
+# NOTE: only applies to cherrypy process ONLY, doesn't apply when started via 'sep' or other processes.
+# uncomment this if you want to do email sending in its own process
+# _default_cherrypy_exclude_categories = ["automated_email_sending"]
+_default_cherrypy_exclude_categories = None
+
+
+class UberSchedulerJob(schedule.Job):
+    def do(self, job_func, *args, **kwargs):
+        # our scheduler needs to not let individual tasks throw exceptions, so wrap it here
+        wrapped_fn = timed(swallow_exceptions(job_func))
+
+        return super().do(wrapped_fn, *args, **kwargs)
+
+
+class UberScheduler(schedule.Scheduler):
+    def every(self, interval=1):
+        # use our custom jobs class so we can override do()
+        job = UberSchedulerJob(interval, self)
+        return job
+
+
+schedule.default_scheduler = UberScheduler()
+
+
+class TimePassed:
+    def __init__(self, minutes_to_wait):
+        assert minutes_to_wait > 0
+        self.minutes_to_wait = minutes_to_wait
+        self.create_time = datetime.utcnow()
+
+    def enough_time_passed(self):
+        assert self.create_time
+        return (datetime.utcnow() - self.create_time).seconds > self.minutes_to_wait*60
+
+
+_startup_time = TimePassed(2)
+
+
+def schedule_N_times_per_day(times_per_day, fn, *args, **kwargs):
+    # slightly hacky.
+    # issue: what we want to say something like "run these every 6 hours starting at midnight"
+    # schedule module only supports saying "run every 6 hours from server startup" so the scheduling
+    # is dependent on the server runtime, and the schedule isn't consistent based on server start time.
+    #
+    # work around this by creating multiple copies of the same task, scheduled for different times.
+    # CAUTION: if tasks are extremely long-running (more than 1 hour each), they can overlap each other
```

Comment:
I would suggest that to be safe against this case, we should probably have each task acquire a lock unique to that task.

While we might thing that nothing could take longer than an hour to run, we have absolutely had email sending take longer than an hour to run, since AWS throttles how many emails we can send per second.  So taking longer than an hour is a real possibility.

At a glance, we don't use this functionality for email sending, so it's probably not as big a deal.  I only brought it up because "background task that takes longer than an hour to run" is absolutely something which has already happened for us and not just a theoretical impossibility.

I believe that ``UberSchedulerJob`` could be modified such that it wrapped the function with an additional decorator which acquired a specific to that function, e.g. we could declare something like

```
task_locks = defaultdict(RLock)
```

and then grab the lock for that particular function, such that each function would get its own lock.